manifests/server.pp

```diff
@@ -1,17 +1,178 @@
+#   Name of default sshd service to disable
+#
+# @param files
+#   File resources for setting up files
+#
+# @param manage_service
+#   Flag of whether to manage sshd service
+#
+# @param port
+#   TCP port that GSI OpenSSH server listens on
+#
+# @param required_packages
+#   List of package names to be installed (OS specific).
+#
+# @param service_name
+#   Name of gsi-sshd service
+#
+# @param subnets
+#   List of network subnets allowed access via GSI OpenSSH server
+#
 # @example
 #   include profile_gsi_ssh::server
-class profile_gsi_ssh::server {
+class profile_gsi_ssh::server (
+  Array[String] $allow_groups,
+  Hash          $config,
+  String        $config_file,
+  Array[String] $deny_groups,
+  String        $disabled_service_name,
+  Hash          $files,
+  Boolean       $manage_service,
+  String        $port,
+  Array[String] $required_packages,
+  String        $service_name,
+  Array[String] $subnets,
+) {
 
   include ::profile_gsi_ssh::client
+  include ::profile_gsi_ssh::server::cert
+  include ::profile_gsi_ssh::server::gridmap
+
+  # PACKAGES
+  ensure_packages( $required_packages, {'ensure' => 'present'} )
+
+  $file_defaults = {
+    ensure  => file,
+    owner   => 'root',
+    group   => 'root',
+    mode    => '0644',
+  }
+  ensure_resources('file', $files, $file_defaults )
+
+  if ( ! empty($disabled_service_name) ) {
+    service { $disabled_service_name :
+      ensure     => stopped,
+      enable     => false,
+      hasstatus  => true,
+      hasrestart => true,
+    }
+  }
+  if ( $manage_service ) {
+    service { $service_name :
+      ensure     => running,
+      enable     => true,
+      hasstatus  => true,
+      hasrestart => true,
+      require    => [
+        Package[ $required_packages ],
+      ],
+    }
+    # SET DEFAULTS TO NOTIFY SERVICE
+    $config_defaults = {
+      'notify' => Service[$service_name] ,
+      'target' => $config_file,
+    }
+  } else {
+    # SET DEFAULTS TO SKIP NOTIFY SERVICE
+    # THE ENSURE => PRESENT IS A DEFAULT, BUT SETTING IT SO THAT WE CAN SET SOME DEFAULT
+    $config_defaults = {
+      'ensure' => present,
+      'target' => $config_file,
+    }
+  }
+
+  # CONFIGURE gsissh sshd_config FILE
+
+  $puppet_file_header = '# This file is managed by Puppet - Changes may be overwritten'
+  exec { "add puppet header to ${config_file}":
+    command => "sed -i '1s/^/${puppet_file_header}\\n/' '${config_file}'",
+    unless  => "grep '${puppet_file_header}' ${config_file}",
+    path    => [ '/bin', '/usr/bin' ],
+  }
+
+  # CONFIGURE PORT
+  sshd_config {
+    'GSI Port' :
+      key   => 'Port',
+      value => $port,
+    ;
+    default:
+      * => $config_defaults,
+    ;
+  }
+
+# Address CANNOT BE SET IN gsisshd's sshd_config
+#  # ADD ALLOW SUBNETS
+#  sshd_config {
+#    'GSI Address' :
+#      key   => 'Address',
+#      value => $subnets.unique.sort.join(','),
+#    ;
+#    default:
+#      * => $config_defaults,
+#    ;
+#  }
```

Comment:
If we want to restrict access within `gsisshd` based on subnet, then we need to move all of our authentication logic into a `Match` block like we do for standard OpenSSH `sshd`. The `Address` parameter only exists for `Match`'s.